library/core/src/slice/mod.rs

```diff
@@ -307,6 +308,65 @@ impl<T> [T] {
         index.get_mut(self)
     }
 
+    /// Returns a pinned reference to an element or subslice depending on the
+    /// type of index (see [`get`]) or `None` if the index is out of bounds.
+    ///
+    /// [`get`]: #method.get
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_get_pin)]
+    /// use std::pin::Pin;
+    ///
+    /// let v = vec![0, 1, 2].into_boxed_slice();
+    /// let pinned = Pin::from(v);
+    ///
+    /// let x: Option<Pin<&i32>> = pinned.as_ref().get_pin_ref(1);
+    /// assert_eq!(&1, &*x.unwrap());
+    /// ```
+    #[unstable(feature = "slice_get_pin", issue = "none")]
+    #[inline]
+    pub fn get_pin_ref<I>(self: Pin<&Self>, index: I) -> Option<Pin<&I::Output>>
+    where
+        I: SliceIndex<Self>,
+    {
+        // SAFETY: `x` is guaranteed to be pinned because it comes from `self`
+        // which is pinned.
+        unsafe { self.get_ref().get(index).map(|x| Pin::new_unchecked(x)) }
+    }
+
+    /// Returns a pinned mutable reference to an element or subslice depending on the
+    /// type of index (see [`get`]) or `None` if the index is out of bounds.
+    ///
+    /// [`get`]: #method.get
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_get_pin)]
+    /// use std::pin::Pin;
+    ///
+    /// let v = vec![0, 1, 2].into_boxed_slice();
+    /// let mut pinned = Pin::from(v);
+    ///
+    /// if let Some(mut elem) = pinned.as_mut().get_pin_mut(1) {
+    ///     elem.set(10);
+    /// }
+    /// assert_eq!(&*pinned, &[0, 10, 2]);
+    /// ```
+    #[unstable(feature = "slice_get_pin", issue = "none")]
+    #[inline]
+    pub fn get_pin_mut<I>(self: Pin<&mut Self>, index: I) -> Option<Pin<&mut I::Output>>
```

Comment:
The existence of this method means that unsafe code needs to be careful when projecting from a `Pin<&mut [T]>` to a `&mut T` (a mutable reference to one of the slice elements). It can still be done soundly, but the `&mut T` cannot be moved out of if the `Pin<&mut [T]>` is ever exposed to safe code (which could call `get_pin_mut`).

I think it would be good to mention this in the docs.

Reply:
That is correct. `Option::as_pin_mut` also has the same issue. We need to add docs to both.